Add kep for coscheduling base on CRD (#42)
kubernetes-sigs/scheduler-plugins: kep/podgroup-coscheduling/README.md

# Coscheduling based on PodGroup CRD

## Table of Contents

<!-- toc -->
- [Coscheduling based on PodGroup CRD](#Coscheduling based on PodGroup CRD)
  - [Table of Contents](#table-of-contents)
  - [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
  - [Use Cases](#use-cases)
  - [Design Details](#design-details)
    - [PodGroup](#podgroup)
    - [Controller](#controller)
    - [Extension points](#extension-points)
      - [QueueSort](#queuesort)
      - [PreFilter](#prefilter)
      - [Permit](#permit)
      - [PostBind](#postbind)
  - [Known Limitations](#Known Limitations)
<!-- /toc -->

## Motivation
Currently, through the default scheduler of Kubernetes, we cannot ensure a group of pods scheduled at the same time . Under some scene, it would waste resources since some pods need work together, like spark, tensorflow and so on . So, podgroup-coscheduling is aimed at solving the issue.

## Goals
1. Base on scheduling framework, implementing the gang scheduling.
2. Define a CRD name PodGroup to help pod scheduling.
 
## Non-Goals

> Sort the job when we submit jobs to a cluster. Currently, we can only do this base on pods.

Author: a general term of batch job

 
## Use Cases
Spark jobs, tensorflow jobs and other pods have to run together.

## Design Details

### PodGroup

> We define a CRD name PodGroup to help schedule, its definition is as follows:

Reviewer: ```suggestion
A CRD named `PodGroup` is proposed with the following API specs:
```

> ```build

Reviewer: ```suggestion
```go
```

// PodGroupSpec represents the template of a pod group.
type PodGroupSpec struct {
	// MinMember defines the minimal number of members/tasks to run the pod group;
	// if there's not enough resources to start all tasks, the scheduler
	// will not start anyone.

> 	MinMember uint32 `json:"minMember"`

Reviewer: `int32` is more conventional. (applies elsewhere)

	// If specified, indicates the PodGroup's priority. "system-node-critical" and
	// "system-cluster-critical" are two special keywords which indicate the
	// highest priorities with the former being the highest priority. Any other

> 	// name must be defined by creating a PriorityClass object with that name.

Reviewer 2: Does PriorityClass  mean another CRD or something else? Can you introduce more about this type?

Author: name of Priority class, same as the default int K8s! But it is not used not, I wonder if I should remove it

	// If not specified, the PodGroup priority will be default or zero if there is no
	// default.
	// +optional
	PriorityClassName string `json:"priorityClassName,omitempty"`

	// MinResources defines the minimal resource of members/tasks to run the pod group;
	// if there's not enough resources to start all tasks, the scheduler
	// will not start anyone.

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Author: pod

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Reviewer: `MinResources v1.ResourceList` is more conventional.

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Reviewer: To be honest, I don't quite get the point of `MinResources`.

The resource request is just one dimension of a PodGroup's constraints, how can you ensure all Pods of a PodGroup can be scheduled successfully by only looking at its resource requests?

> pod

but the comments said it's for all tasks.

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Author: Yes, this seems to be not necessary. Actually we did not set it in our current implementation, the value is achieved from Pod. I would remove it.

	// MaxScheduleTime defines the maximal time of members/tasks to wait before run the pod group;

> 	MaxScheduleTime *metav1.Duration `json:"maxScheduleTime,omitempty"`

Reviewer: We should avoid defining API as Go type `Duration`, use `int64` instead, and rename it like `ScheduleTimeoutSeconds` or `WaitTimeoutSeconds`.

Ref:

- [api-conventions.md#units](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units)
- [api-conventions.md#naming-conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions)

}

// PodGroupStatus represents the current state of a pod group.
type PodGroupStatus struct {
	// Current phase of PodGroup.
	Phase PodGroupPhase `json:"phase"`

	// OccupiedBy marks the podgroup occupied by which group.
	// Owner reference would be used to filled it, it not initialize, it is empty

> 	OccupiedBy string `json:"occupiedBy,omitempty"`

Reviewer: Are you trying to describe a hierarchical relation here? For example, a PodGroup is owned/referred by another PodGroup? If so, you can inject the relation into`ObjectMeta.OwnerReferences`, so don't need to define a new API field.

Author: Not, in our current implement this is just UID of a deployment, statefulSet or Other CR, e.g tf job or mpi job.

	// The number of actively running pods.
	// +optional
	Scheduled uint32 `json:"scheduled"`

	// The number of actively running pods.
	// +optional
	Running uint32 `json:"running"`

	// The number of pods which reached phase Succeeded.
	// +optional
	Succeeded uint32 `json:"succeeded"`

	// The number of pods which reached phase Failed.
	// +optional
	Failed uint32 `json:"failed"`

	// ScheduleStartTime of the group
	ScheduleStartTime metav1.Time `json:"scheduleStartTime"`
}
```

### Controller

> We define a controller to reconcile PodGroup status, and we can query the job status through describe the PodGroup. Any pod in a group failed, the Group Status is marked Failed. Controller would also help whem recovering from abnormal cases, e.g. batch scheduling is interpreted due to

Reviewer: - ... through describing ...
- ... Once a pod in a group failed ... is marked as Failed...
- ... Controller would also help to recover from ... 
- interpreted -> interrupted?

cluster upgrade.

### Extension points

#### QueueSort

To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:

1. Sort based on Pod Priority
2. Pod does not belong to any group
3. PodGroup creation time
4. Pod creation time

#### PreFilter

> This extension helps pre-filter pods. It is useful, especially when there are not enough resources in a cluster. The main proggress is as follows:

Reviewer: ```suggestion
This extension helps pre-filter pods. It is useful, especially when there are not enough resources in a cluster. The overall flow works as below:
```

> 1. Allow the pods do not belong to any group.

Reviewer: ```suggestion
1. Allow the pods that do not belong to any group.
```

2. If there are no groups scheduling, we check resource, if enough, we allow the pod.

> 3. If there are groups running, we check if the current pod belong the max finished group, if it is, we allow it.

Reviewer 3: What‘s the `Max finished group`？
@cwdsuzhou

Author: The group who has the max progress.

Reviewer: Please explain `max` in the document.
